packages/react-dom-bindings/src/server/ReactDOMServerFormatConfig.js

```diff
@@ -2234,54 +2267,98 @@ const styleInsertionFunction =
 const clientRenderFunction =
   'function $RX(b,c,d,e){var a=document.getElementById(b);a&&(b=a.previousSibling,b.data="$!",a=a.dataset,c&&(a.dgst=c),d&&(a.msg=d),e&&(a.stck=e),b._reactRetry&&b._reactRetry())}';
 
-const completeSegmentScript1Full = stringToPrecomputedChunk(
+function assertIsScriptExecution(
+  s: InstructionWriterState,
+): ScriptExecutionWriterState {
+  if (s.kind !== 'ScriptExecution') {
+    // This should be unreachable, since running with enableFizzNoScriptExecution=false
+    // means React will always use ScriptExecutionWriterState
+    throw new Error(
+      'Unexpected data writer when enableFizzNoScriptExecution=false. This is a bug in React. Please file an issue.',
+    );
+  }
+  return s;
+}
+
+const completeSegmentScriptFullStart = stringToPrecomputedChunk(
   completeSegmentFunction + ';$RS("',
 );
-const completeSegmentScript1Partial = stringToPrecomputedChunk('$RS("');
-const completeSegmentScript2 = stringToPrecomputedChunk('","');
-const completeSegmentScript3 = stringToPrecomputedChunk('")</script>');
+const completeSegmentScriptPartialStart = stringToPrecomputedChunk('$RS("');
+const completeSegmentInterstitial = stringToPrecomputedChunk('","');
+const completeSegmentEnd = stringToPrecomputedChunk('"');
+
+const completeSegmentScriptEnd = stringToPrecomputedChunk(')</script>');
+
+const completeSegmentDataStart = stringToPrecomputedChunk('"$RS","');
 
 export function writeCompletedSegmentInstruction(
   destination: Destination,
   responseState: ResponseState,
   contentSegmentID: number,
 ): boolean {
-  writeChunk(destination, responseState.startInlineScript);
-  if (!responseState.sentCompleteSegmentFunction) {
-    // The first time we write this, we'll need to include the full implementation.
-    responseState.sentCompleteSegmentFunction = true;
-    writeChunk(destination, completeSegmentScript1Full);
+  const writerState = responseState.writerState;
+  if (!enableFizzNoScriptExecution || writerState.kind === 'ScriptExecution') {
+    const scriptState = assertIsScriptExecution(writerState);
+    // <script ...>$RS("
+    // <script ...>function $RS(...) {...}; $RS("
+    writeChunk(destination, scriptState.startInlineScript);
+    if (!scriptState.sentCompleteSegmentFunction) {
+      // The first time we write this, we'll need to include the full implementation.
+      scriptState.sentCompleteSegmentFunction = true;
+      writeChunk(destination, completeSegmentScriptFullStart);
+    } else {
+      // Future calls can just reuse the same function.
+      writeChunk(destination, completeSegmentScriptPartialStart);
+    }
   } else {
-    // Future calls can just reuse the same function.
-    writeChunk(destination, completeSegmentScript1Partial);
+    // <div data-attr='["$RS", "
+    (writerState: ScriptDataWriterState);
+    writeChunk(destination, dataElementStart);
```

Comment:
Just noting for myself and others...

Originally I was hoping that we could attach these attributes to the node added by `writeStartSegment`. However, I don't think that would work.

```
<div hidden data-attr="...">
  <div>first</div>
  <div>second</div>
</div>
```

In this case the mutation observer would fire when the first div was added to the tree. Whether or not the second two divs have been streamed in yet. There's also no signal when it is done.

Therefore, unfortunately I think the best we can do is emit an extra node after - like this.

Reply:
For segments completing, I think we could maybe listen to any element being added to the segment and just move them as they come in.

For boundaries completing, we'd still need a separate element regardless to know when it's safe to do the reveal.

Probably not worth the complexity to special case segments completing.